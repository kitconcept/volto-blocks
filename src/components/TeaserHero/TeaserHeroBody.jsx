import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { Message } from 'semantic-ui-react';
import { defineMessages, injectIntl } from 'react-intl';
import cx from 'classnames';
import find from 'lodash/find';
import { getContent } from '@plone/volto/actions';
import { flattenToAppURL } from '@plone/volto/helpers';
import templates from './templates';

const messages = defineMessages({
  PleaseChooseContent: {
    id: 'Please choose an existing content as source for this element',
    defaultMessage:
      'Please choose an existing content as source for this element',
  },
});

const TeaserHeroBody = ({ data, id, isEditMode, intl }) => {
  const contentSubrequests = useSelector(state => state.content.subrequests);
  const dispatch = useDispatch();
  const result = contentSubrequests?.[id]?.data;

  React.useEffect(() => {
    if (data.href) {
      dispatch(getContent(data.href, null, id));
    }
  }, [dispatch, data, id]);

  return (
    <>
      {!data.href && (
        <Message>
          <div className="teaser-item default">
            <img
              src={
                find(templates(), item => item.variation === data.variation)
                  ?.image
              }
              alt=""
            />
            <p>{intl.formatMessage(messages.PleaseChooseContent)}</p>
          </div>
        </Message>
      )}
      {data.href && result && (
        <div className={cx('teaser-item', data.variation, {})}>
          {(() => {
            const item = (
              <>
                {result?.preview_image && (
                  <img
                    src={flattenToAppURL(result.preview_image.download)}
                    alt=""
                  />
                )}
                <div>
                  <h3>{result.title}</h3>
                  <p>{result.description}</p>
                </div>
              </>
            );
            if (!isEditMode) {
              return (
                <Link
                  to={flattenToAppURL(result['@id'])}
                  target={data.openLinkInNewTab ? '_blank' : null}
                >
                  {item}
                </Link>
              );
            } else {
              return item;
            }
          })()}
        </div>
      )}
    </>
  );
};

TeaserHeroBody.propTypes = {
  data: PropTypes.objectOf(PropTypes.any).isRequired,
  isEditMode: PropTypes.bool,
};

export default injectIntl(TeaserHeroBody);
