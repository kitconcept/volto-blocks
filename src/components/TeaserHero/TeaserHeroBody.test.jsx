import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import { MemoryRouter } from 'react-router-dom';

import TeaserHeroBody from './TeaserHeroBody';

const mockStore = configureStore();

test('renders a Grid Tile Proxy Item component', () => {
  const store = mockStore({
    content: {
      subrequests: {
        '1c58b5a2-26fb-4112-832b-fad88af1ae0e': {
          data: {
            title: 'the title',
            description: 'the description',
            image: { download: 'http://image' },
            '@id': 'http://theitem',
          },
        },
      },
    },
    intl: {
      locale: 'en',
      messages: {},
    },
  });
  const component = renderer.create(
    <Provider store={store}>
      <MemoryRouter>
        <TeaserHeroBody
          id="1c58b5a2-26fb-4112-832b-fad88af1ae0e"
          data={{
            '@type': 'teaserHero',
            href: '/de/a-page/news1',
            variation: 'left',
            padded: true,
          }}
          tile="1c58b5a2-26fb-4112-832b-fad88af1ae0e"
          onChangeTile={() => {}}
        />
      </MemoryRouter>
    </Provider>,
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
