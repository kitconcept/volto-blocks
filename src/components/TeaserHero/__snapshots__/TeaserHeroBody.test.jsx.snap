// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Grid Block Proxy Item component 1`] = `
<div
  className="teaser-item left"
>
  <a
    href="/http://theitem"
    onClick={[Function]}
    target={null}
  >
    <img
      alt=""
      src="http://image"
    />
    <div>
      <h3>
        the title
      </h3>
      <p>
        the description
      </p>
    </div>
  </a>
</div>
`;
